### PR TITLE
Port Windows PowerShell AppLocker and DeviceGuard UMCI application white listing support

### PR DESCRIPTION
**Pull request by PaulHigin**
## PR Summary

These changes port Windows PowerShell support for Applocker and DeviceGuard User Mode Code Integrity (UMCI) to PSCore6.  Windows PowerShell uses public APIs to determine if a system is in locked down mode via AppLocker or DeviceGuard, and automatically runs in constrained language mode.  For more information about PowerShell constrained language, see: https://blogs.msdn.microsoft.com/powershell/2017/11/02/powershell-constrained-language-mode/

This support for application white listing has mostly existed in PSCore6, but the primary APIs were stubbed out in CorePSStub.cs because they relied on Windows only DeviceGuard (wldp.dll) and AppLocker (Safer APIs) public APIs.  These changes re-implement PowerShell lock down APIs on PSCore6 for Windows platforms only.  The AppLocker and DeviceGuard public APIs are currently only implemented in Windows OSes and are not supported on Linux or MacOS platforms.

Tests have also been ported to PSCore6 and run only for Windows platforms.

## PR Checklist

Note: Please mark anything not applicable to this PR `NA`.

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [x] Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [`NA`] User facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [ ] Issue filed - Issue link:
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
    - [x] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.

### CODE REVIEW

File: test/powershell/Modules/Microsoft.PowerShell.Security/AmsiInterface.Tests.ps1
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Security/AmsiInterface.Tests.ps1]: Please add EOL

File: src/System.Management.Automation/security/wldpNativeMethods.cs
Comment (adityapatwardhan) [inline src/System.Management.Automation/security/wldpNativeMethods.cs]: Please add EOL.

**Review by adityapatwardhan — APPROVED** [review]

**Comment by TravisEz13** [comment]
`[feature]` goes in the commit description not the PR title.

**Comment by PaulHigin** [comment]
Thanks, I forgot that.  Will submit another commit soon.

File: test/powershell/Modules/Microsoft.PowerShell.Security/AmsiInterface.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Security/AmsiInterface.Tests.ps1]: For consistency (in case we ever need to update the header), can you use the exact form of this header and ensure it's the top two lines first.

File: test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1]: Use standard header here, too

File: test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1]: BeExactly

File: test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1]: Can this be changed to use ShouldBeErrorId pattern?  Same for all the exception validation below.

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Security/AmsiInterface.Tests.ps1
Comment (PaulHigin) [reply test/powershell/Modules/Microsoft.PowerShell.Security/AmsiInterface.Tests.ps1]: Done

File: src/System.Management.Automation/security/wldpNativeMethods.cs
Comment (PaulHigin) [reply src/System.Management.Automation/security/wldpNativeMethods.cs]: Done

File: test/powershell/Modules/Microsoft.PowerShell.Security/AmsiInterface.Tests.ps1
Comment (PaulHigin) [reply test/powershell/Modules/Microsoft.PowerShell.Security/AmsiInterface.Tests.ps1]: Done

File: test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1
Comment (PaulHigin) [reply test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1]: Done

File: test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1
Comment (PaulHigin) [reply test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1]: Done

File: test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1
Comment (PaulHigin) [reply test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1]: No

**Review by SteveL-MSFT — APPROVED** [review]

File: src/System.Management.Automation/security/wldpNativeMethods.cs
Comment (TravisEz13) [inline src/System.Management.Automation/security/wldpNativeMethods.cs]: Can you make sure that the copyright header is applied?

**Review by TravisEz13 — DISMISSED** [review]

File: src/System.Management.Automation/security/wldpNativeMethods.cs
Comment (PaulHigin) [reply src/System.Management.Automation/security/wldpNativeMethods.cs]: Done

**Comment by TravisEz13** [comment]
Let's run compliance tests on this before merging.

File: src/System.Management.Automation/CoreCLR/CorePsStub.cs
Comment (iSazonov) [inline src/System.Management.Automation/CoreCLR/CorePsStub.cs]: Please add final dot in the _public_ comment.

File: test/powershell/Modules/Microsoft.PowerShell.Security/AmsiInterface.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Security/AmsiInterface.Tests.ps1]: Why we don't use `ShouldBeErrorId`?

File: src/System.Management.Automation/CoreCLR/CorePsStub.cs
Comment (PaulHigin) [reply src/System.Management.Automation/CoreCLR/CorePsStub.cs]: Will do.

File: test/powershell/Modules/Microsoft.PowerShell.Security/AmsiInterface.Tests.ps1
Comment (PaulHigin) [reply test/powershell/Modules/Microsoft.PowerShell.Security/AmsiInterface.Tests.ps1]: The tests were already like this.  I feel changing all tests is outside the scope of this change.  Please create a new issue.

Comment (TravisEz13) [reply test/powershell/Modules/Microsoft.PowerShell.Security/AmsiInterface.Tests.ps1]: Can one of you file an issue to fix the test?

**Comment by TravisEz13** [comment]
Please fix the merge conflicts

File: test/powershell/Modules/Microsoft.PowerShell.Security/AmsiInterface.Tests.ps1
Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Security/AmsiInterface.Tests.ps1]: Done #6151

File: test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1
Comment (TravisEz13) [inline test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1]: copyright should be first line

**Review by TravisEz13 — APPROVED** [review]
one minor comment

File: test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1
Comment (TravisEz13) [reply test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1]: @PaulHigin  Can you address this?

**Comment by PaulHigin** [comment]
@SteveL-MSFT  Why is this PR being abandoned?

**Comment by SteveL-MSFT** [comment]
@PaulHigin it's not abandoned (yet).  We enabled a new bot to automatically flag stale PRs.  For this PR, I think you just need to resolve the merge conflicts and we can proceed.

File: test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1
Comment (TravisEz13) [reply test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1]: @PaulHigin  Can you fix this?

**Comment by PaulHigin** [comment]
@TravisEz13  I just noticed that this PR is still active.  Is there any reason why it has not been merged?

**Comment by TravisEz13** [comment]
@PaulHigin  Please address my comment: https://github.com/PowerShell/PowerShell/pull/6133#pullrequestreview-96678613

File: test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1
Comment (PaulHigin) [reply test/powershell/Modules/Microsoft.PowerShell.Security/ConstrainedLanguageRestriction.Tests.ps1]: @TravisEz13  ?? The copyright is the first line.  AFAIK this was fixed a month ago.  In any case I don't see why the check-in should be held up for month about this...

**Comment by TravisEz13** [comment]
restarting windows tests due to the following failure
```
Describing Implicit remoting on restricted ISS
VERBOSE: Using Global AppVeyor Credential
pwsh.exe : New-PSSession : [localhost] Connecting to remote server localhost failed with the following error message : Access is denied. For more information, see the about_Remote_Troubleshooting Help topic.
At C:\projects\powershell\build.psm1:1224 char:48
+ ... art-NativeExecution -sb {& $powershell -noprofile -c $command} | ForE ...
+                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (New-PSSession :...ing Help topic.:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
At C:\projects\powershell\test\tools\Modules\HelpersRemoting\HelpersRemoting.psm1:96 char:20
+         $session = New-PSSession @parameters
+                    ~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : OpenError: (System.Manageme....RemoteRunspace:RemoteRunspace) [New-PSSession], PSRemotingTransportException
+ FullyQualifiedErrorId : AccessDenied,PSSessionOpenFailed
    [-] Error occurred in Describe block 19.89s
      Expected a value, but got $null or empty.
      1500:             $session | Should -Not -BeNullOrEmpty
      at <ScriptBlock>, C:\projects\powershell\test\powershell\Modules\Microsoft.PowerShell.Utility\Implicit.Remoting.Tests.ps1: line 1500
      at Invoke-Blocks, C:\projects\powershell\src\powershell-win-core\bin\Release\netcoreapp2.0\win7-x64\publish\Modules\Pester\4.3.1\Functions\SetupTeardown.ps1: line 140
      at Invoke-TestGroupSetupBlocks, C:\projects\powershell\src\powershell-win-core\bin\Release\netcoreapp2.0\win7-x64\publish\Modules\Pester\4.3.1\Functions\SetupTeardown.ps1: line 125
      at DescribeImpl, C:\projects\powershell\src\powershell-win-core\bin\Release\netcoreapp2.0\win7-x64\publish\Modules\Pester\4.3.1\Functions\Describe.ps1: line 157
  Describing Implicit remoting tests
VERBOSE: Using Global AppVeyor Credential
    Context Get-Command <Imported-Module> and <Imported-Module.Name> work (Windows 7: #334112)
```

**Comment by TravisEz13** [comment]
@PaulHigin It failed again with the same error.

**Comment by PaulHigin** [comment]
@TravisEz13  These are remoting tests that have nothing to do with these changes.  The fact that they are failing should not affect this PR.

**Comment by TravisEz13** [comment]
We don't merge with CI tests failing, except if the PR is required to unblock. I've restarted the AppVeyor tests.

**Pull request merged**
